### DISCUSSION TITLE
Errors when trying to make server-side filtering:

### DISCUSSION THREAD

**Discussion by BatuKhurtsilava** [supabase/supabase#19068] [Questions] [answered]
query.ts:446  TypeError: query.range is not a function     at getCategoryData (useGetData.tsx:40:1),  query.ts:446  TypeError: query.eq is not a function     at getCategoryData (useGetData.tsx:40:1)

import React from "react";
import supabase from "../services/supabase";
import { useQuery } from "@tanstack/react-query";
import { useSearchParams } from "react-router-dom";
import { PRODS_QUANTITY_PAGE } from "../utils/constants";
import { useQueryClient } from "@tanstack/react-query";

interface FilterObject {
  filterByPrice: string | null;
  filterByStock: string | null;
  filterByBrand: string | null;
  filterBySale: string | null;
}

async function getCategoryData(
  category: string,
  filterParamObject?: FilterObject,
  page?: number | null
) {
  const priceRange = filterParamObject?.filterByPrice?.split("-");
  const stockRange = filterParamObject?.filterByStock?.split("-");

  let query: any = await supabase
    .from(category)
    .select("*", { count: "exact" });

  if (filterParamObject!.filterByBrand !== null)
    query = query.eq("Brand", filterParamObject!.filterByBrand);
  if (filterParamObject!.filterByPrice !== null)
    query = query.gte("Price", priceRange![0]).lte("Price", priceRange![1]);
  if (filterParamObject!.filterByStock !== null)
    query = query.gte("Stock", stockRange![0]).lte("Stock", stockRange![1]);
  if (filterParamObject!.filterBySale === "Without Discount")
    query = query.is("Sale", null);
  if (filterParamObject!.filterBySale === "With Discount")
    query = query.neq("Sale", null);
  if (page) {
    const from = (page! - 1) * PRODS_QUANTITY_PAGE;
    const to = from + PRODS_QUANTITY_PAGE - 1;
    query = query.range(from, to);
  }

  const { data, error, count } = await query;
  if (error) {
    console.log(error);
    throw new Error(`Could not load ${category}  data`);
  }

  console.log(data);

  return { data, count };
}

export const useGetData = (category: string) => {
  const [searchParams] = useSearchParams();
  const queryClient = useQueryClient();
  const page = !searchParams.get("page") ? 1 : Number(searchParams.get("page"));
  const filterParamObject = {
    filterByBrand: searchParams.get("Brand"),
    filterByPrice: searchParams.get("Price"),
    filterBySale: searchParams.get("Sale"),
    filterByStock: searchParams.get("Stock"),
  };

  const {
    isLoading,
    data: { data: products, count } = {},
    error,
  } = useQuery({
    queryKey: [
      category,
      page,
      filterParamObject.filterByBrand,
      filterParamObject.filterByPrice,
      filterParamObject.filterBySale,
      filterParamObject.filterByStock,
    ],
    queryFn: () => getCategoryData(category, filterParamObject, page),
  });

  const pageCount = Math.ceil(count! / PRODS_QUANTITY_PAGE);

  if (page < pageCount)
    queryClient.prefetchQuery({
      queryKey: [
        category,
        page,
        filterParamObject.filterByBrand,
        filterParamObject.filterByPrice,
        filterParamObject.filterBySale,
        filterParamObject.filterByStock,
      ],
      queryFn: () => getCategoryData(category, filterParamObject, page + 1),
    });

  if (page > 1)
    queryClient.prefetchQuery({
      queryKey: [
        category,
        page,
        filterParamObject.filterByBrand,
        filterParamObject.filterByPrice,
        filterParamObject.filterBySale,
        filterParamObject.filterByStock,
      ],
      queryFn: () => getCategoryData(category, filterParamObject, page - 1),
    });
  return { isLoading, products, count, error };
};

**Comment by GaryAustin1** [comment, marked as answer]
It helps to enclose code in `'''js   my code '''` to make it readable.
https://supabase.com/docs/reference/javascript/using-filters?example=conditional-chaining
![image](https://github.com/supabase/supabase/assets/54564956/d76f760d-1203-4bcf-9271-946934dde7b5)

You are not setting things correctly.  As soon as you do await it executes.  So you need to build it up as shown.

**Reply by BatuKhurtsilava** [reply]
Thank you so  much  my friend
